optax/transforms/_freezing_test.py

```diff
@@ -0,0 +1,283 @@
+                    - GRAD_NESTED['layer1'][1]['bias']}
+         ],
+         'layer2': {
+           'w': PARAMS_NESTED['layer2']['w'] - GRAD_NESTED['layer2']['w'],
+           'b': PARAMS_NESTED['layer2']['b'] - GRAD_NESTED['layer2']['b'],
+         },
+         'misc': (PARAMS_NESTED['misc'][0] - GRAD_NESTED['misc'][0],),
+       }
+      ),
+      # nested: freeze entire layer2
+      ('nested_freeze_layer2',
+       PARAMS_NESTED, GRAD_NESTED,
+       {
+         'layer1': [False, {'bias': False}],
+         'layer2': {'w': True,  'b': True},
+         'misc':   (False,),
+       },
+       {
+         'layer1': [
+           PARAMS_NESTED['layer1'][0] - GRAD_NESTED['layer1'][0],
+           {'bias': PARAMS_NESTED['layer1'][1]['bias']
+                    - GRAD_NESTED['layer1'][1]['bias']}
+         ],
+         'layer2': PARAMS_NESTED['layer2'],  # frozen
+         'misc':   (PARAMS_NESTED['misc'][0] - GRAD_NESTED['misc'][0],),
+       }
+      ),
+  ])
+  def test_selective_optimizer_effect(
+      self, params, grads, freeze_mask, expected_params):
+    """Tests that selective_optimizer only updates the unfrozen leaves."""
+    inner_opt = alias.sgd(learning_rate=1.0)
+    optimizer = _freezing.selective_optimizer(inner_opt, freeze_mask)
+
+    state = optimizer.init(params)
+    updates, _ = optimizer.update(grads, state, params)
+    new_params = update.apply_updates(params, updates)
+
+    chex.assert_trees_all_close(new_params, expected_params, atol=1e-6)
+
+
+class FreezeExtraTest(parameterized.TestCase):
+
+  def test_nested_freeze_all(self):
+    mask = {
+      'layer1': [True, {'bias': True}],
+      'layer2': {'w': True, 'b': True},
+      'misc': (True,)
+    }
+    opt = _freezing.freeze(mask)
+    state = opt.init(PARAMS_NESTED)
+    updates, _ = opt.update(GRAD_NESTED, state, PARAMS_NESTED)
+    chex.assert_trees_all_close(
+        updates,
+        jax.tree.map(lambda g: g * 0, GRAD_NESTED),
+        atol=0
+    )
+
+  def test_nested_freeze_none(self):
+    mask = {
+      'layer1': [False, {'bias': False}],
+      'layer2': {'w': False, 'b': False},
+      'misc': (False,)
+    }
+    opt = _freezing.freeze(mask)
+    state = opt.init(PARAMS_NESTED)
+    updates, _ = opt.update(GRAD_NESTED, state, PARAMS_NESTED)
+    chex.assert_trees_all_close(updates, GRAD_NESTED, atol=0)
+
+  @parameterized.named_parameters([
+    ('py_bool', True),
+    ('jax_bool', jnp.array(False)),
+  ])
+  def test_scalar_bool_broadcast(self, scalar_mask):
+    opt = _freezing.freeze(scalar_mask)
+    state = opt.init(PARAMS_FLAT)
+    updates, _ = opt.update(GRAD_FLAT, state, PARAMS_FLAT)
+    expected = (
+      jax.tree.map(jnp.zeros_like, PARAMS_FLAT)
+      if bool(scalar_mask) else GRAD_FLAT
+    )
+    chex.assert_trees_all_close(updates, expected, atol=0)
+
+  def test_bad_structure_raises(self):
+    bad_mask = {'layer1': [True]}  # missing the bias leaf
+    opt = _freezing.freeze(bad_mask)
+    with self.assertRaises(Exception):

```

Comment:
Can you make the exception more specific and check for a specific message using the more specific: `self.assertRaisesRegex` please?

[https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaisesRegex)